### PR TITLE
Add Data Volume documentation

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
This PR adds documentation explaining the benefits of Data Volumes and gives some examples.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
None

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: doc/datavolumes.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+4. Wait for the importer pod to finish.
+5. Start the VMs using the imported disk.
+There is no mechanism to stop 5 from happening before the import is complete, so once can attempt to start the VM before the disk has been completely imported, with obvious bad results.
+
+Now lets do the same process but using DVs.
+1. Create a VM definition that references a DV template, which includes the external URL that contains the disk image.
+2. A DV is created from the template that in turn creates an underlying PVC with the correct annotation.
+3. The importer pod is created like before.
+4. Until the DV status is complete, the virt launcher controller will not schedule the VM to be launched.
+We now have a fully controller mechanism where we can define a VM using a DV with a disk image from an external source, that cannot be scheduled to run until the import has been completed.
```
Comment (j-griffith) [inline doc/datavolumes.md]: s/controller/controlled/  ?

Comment (awels) [reply doc/datavolumes.md]: Fixed

File: doc/datavolumes.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
+# Data Volumes
+
+## Introduction
+Data Volumes(DV) are an abstraction on top of Persistent Volume Claims(PVC) and the Containerized Data Importer(CDI). The DV will monitor and orchestrate the upload/import of the data into the PVC. Once the process is completed, the DV will be in a consistent state that allow consumers to make certain assumptions about the DV in order to progress their own orchestration.
+
+Why is this an improvement over simply looking at the state annotation created and managed by CDI? For one, CDI has different state annotations for different operations (import vs cloning vs upload), where DVs have a single state field. This makes it a lot easier to check that one field vs trying to map all the different annotations in your process. Also DVs will always have a status field, where PVCs may or may not have a status annotation, so you will have to account for that as well.
```
Comment (mhenriks) [inline doc/datavolumes.md]: Actually, as of the controller dedupe, import/clone/upload all use the same annotation "cdi.kubevirt.io/storage.pod.phase".  Not sure this was actually a good idea.

The datavolume phases are a little more descriptive though. 

To me, the real (only?) value of DataVolume is the additional orchestration provided when used in conjunction with KubeVirt.

File: doc/datavolumes.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      requests:
+        storage: "128Mi"
+```
+[Get example](example-clone-dv.yaml)
+
+## Upload Data Volumes
+TBD, this has not been implemented yet.
+
+## Kubevirt integration
+Kubevirt is an extension to Kubernetes that allows one to run Virtual Machines(VM) on the same infra structure as the containers managed by Kubernetes. CDI provides a mechanism to get a disk image into a PVC in order for Kubevirt to consume it. The following steps have to be taken in order for Kubevirt to consume a CDI provided disk image.
```
Comment (mhenriks) [inline doc/datavolumes.md]: Maybe link to KubeVirt here?

File: doc/example-clone-dv.yaml [hunk 4]
```diff
@@ -0,0 +1,15 @@
+apiVersion: cdi.kubevirt.io/v1alpha1
```
Comment (mhenriks) [inline doc/example-clone-dv.yaml]: should the manifests go in the manifests/example directory?

File: doc/datavolumes.md (hunk 2 above)
Comment (awels) [reply doc/datavolumes.md]: Well the virt launcher controller uses the phase to determine if we can launch a VM, nothing stopping it from looking at the annotation. So now having a harder time spotting the benefits of the DV.

Comment (davidvossel) [reply doc/datavolumes.md]: The benefit of DataVolumes is it provides a versioned API projects like KubeVirt can integrate with.

File: doc/datavolumes.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+* 'Blank': No status available.
+* Pending: The operation is pending, but has not been scheduled yet.
+* PVCBound: The PVC associated with the operation has been bound.
+* Import/Clone/UploadScheduled: The operation (import/clone/upload) has been scheduled.
+* Import/Clone/UploadInProgress: The operation (import/clone/upload) is in progress.
+* Succeeded: The operation has succeeded.
+* Failed: The operation has failed.
+* Unknown: Unknown status.
+
+## Import Data Volumes
```
Comment (davidvossel) [inline doc/datavolumes.md]: Maybe Title the "HTTP Source" and the clone one as "PVC Source"

Behind the scenes we know there are import and clone controllers involved, but from a DataVolume API perspective there are only sources.

File: doc/datavolumes.md (hunk 2 above)
Comment (davidvossel) [reply doc/datavolumes.md]: To clarify, I wouldn't try to justify DataVolumes outside of KubeVirt integration.

File: doc/datavolumes.md (hunk 5 above)
Comment (awels) [reply doc/datavolumes.md]: Yeah sounds good, will change.

File: doc/datavolumes.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+```
+[Get example](../manifests/example/example-clone-dv.yaml)
+
+## Upload Data Volumes
+TBD, this has not been implemented yet.
+
+## Kubevirt integration
+[Kubevirt](https://github.com/kubevirt/kubevirt) is an extension to Kubernetes that allows one to run Virtual Machines(VM) on the same infra structure as the containers managed by Kubernetes. CDI provides a mechanism to get a disk image into a PVC in order for Kubevirt to consume it. The following steps have to be taken in order for Kubevirt to consume a CDI provided disk image.
+1. Create a PVC with an annotation to for instance import from an external URL.
+2. The PVC is created and an importer pod is start that attempts to get the image from the external source.
```
[the remark this answers is not in the document]
Comment (awels) [reply doc/datavolumes.md]: Well that is not entirely correct, With DataVolume you create a VM definition that has a DV template, from that template a DV is created, which in turn creates a PVC. Which is what I spell out.

File: doc/datavolumes.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+```yaml
+apiVersion: cdi.kubevirt.io/v1alpha1
+kind: DataVolume
+metadata:
+  name: "example-import-dv"
+spec:
+  source:
+      http:
+         url: "https://download.cirros-cloud.net/0.4.0/cirros-0.4.0-x86_64-disk.img" # Or S3
+         secretRef: "" # Optional
```
Comment (aglitke) [inline doc/datavolumes.md]: What is the expected format of the secret you reference?  Does that secret need to contain specific named fields?

Comment (j-griffith) [reply doc/datavolumes.md]: Yeah, would be really good if we had an example for this

Comment (awels) [reply doc/datavolumes.md]: Added some links to an example secret.

**Pull request merged**